datafusion/core/src/physical_optimizer/limit_aggregation.rs

```diff
@@ -0,0 +1,67 @@
+use std::sync::Arc;
+use datafusion_common::config::ConfigOptions;
+use crate::physical_optimizer::PhysicalOptimizerRule;
+use crate::physical_plan::ExecutionPlan;
+use crate::physical_plan::sorts::sort::SortExec;
+use datafusion_common::{DataFusionError, Result};
+use crate::physical_plan::aggregates::AggregateExec;
+
+pub struct LimitAggregation {
+
+}
+
+impl LimitAggregation {
+    pub fn new() -> Self {
+        Self {}
+    }
+
+    fn recurse(plan: Arc<dyn ExecutionPlan>) -> Result<Arc<dyn ExecutionPlan>> {
+        if let Some(sort) = plan.as_any().downcast_ref::<SortExec>() {
+            let children = sort.children();
+            let child = match children.as_slice() {
+                [] => Err(DataFusionError::Execution("Sorts should have children".to_string()))?,
+                [child] => child,
+                _ => Err(DataFusionError::Execution("Sorts should have 1 child".to_string()))?,
+            };
+            let binding = (*child).as_any();
+            if let Some(aggr) = binding.downcast_ref::<AggregateExec>() {
```

Comment:
I think we should only do this for accumulators like min/max combined with respectively sorting asc/desc that give correct results under this optimization.
Are there other aggregation/sort combinations we expect?

Reply:
Ah I see it was added in `AggregateExec` - is it possible to perform the check in the rule?

Reply:
> only do this for accumulators like min/max
> Are there other aggregation/sort combinations we expect?

I've thought about it for a while, and I'm fairly certain this is a (valuable but) specialized optimization that is only valid for min/max.

Reply:
Yeah, I think for this to work you either need something like min/max where one partition "wins" when the partial states are merged or you need to have the total order of groups consistent between partitions. I'm sure you could come up with an aggregation function which has the latter property (or have the data partitioned a certain way in which it would be true for standard aggregation functions) but in practice it's probably not the case.

Reply:
I filed https://github.com/apache/arrow-datafusion/issues/7198 to try and describe what the end usecase is so maybe that can help focus this conversation

I also think it might end up being a very special operator for MIN/MAX w/ limit

@comphead  had the following pointer https://github.com/apache/arrow-datafusion/issues/6899#issuecomment-1631660067 for prior art

> Spark does the similar way: it sorts and limits data per partition then sends the output to single partition where final sort/limit performed. Spark has the logic encapsulated in separate operator and looks like https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala#L311
> it also contains optimization like if tuples already ordered it will skip the excessive ordering, the same happens for projection

Reply:
Resolving this conversation since it appears to be about when to (correctly) invoke this optimization in the query plan, and I think my latest push solves that. Please re-open a new discussion if I'm wrong!